check-mysql-replication-status: Lag flapping protection

DrMurx opened the pull request:
## Pull Request Checklist

#### General

- [x] Update Changelog following the conventions laid out [here](https://github.com/sensu-plugins/community/blob/master/HOW_WE_CHANGELOG.md)

- [x] Update README with any necessary configuration snippets

- [x] RuboCop passes

- [ ] Existing tests pass

- [x] First test written 😄 

#### Purpose

MariaDB/MySQL sometimes wrongly reports a very high replication lag for a short moment. Flapping protection helps mitigating this issue better than setting `occurrences` in sensu's `checks` definition because you don't lose any alerting granularity.

I had to do major refactoring on `check-mysql-replication-status` to allow testing.

While doing so, I discovered two minor flaws. I've fixed one of them in dedicated commit upfront so that they could be cherry-picked independently if needed.

The other flaw affects the warning/critical thresholds which are exclusive, contrary to the common practice in other checks. Fixing this would be a breaking change, though, so I left a comment in code for a future release.

#### Known Compatibility Issues

DrMurx commented:
@majormoses The `codeclimate-test-reporter` in the (existing) `test/spec_helper.rb` prevents the tests from running. Can I ditch it?

majormoses commented:
> The codeclimate-test-reporter in the (existing) test/spec_helper.rb prevents the tests from running. Can I ditch it?

ya we kinda abandoned that as we had pretty low unit test coverage and I have decided to focus more on integration testing which can't really be measured in the same way. We can drop it and bring it back later if we get enough coverage to see the value in it.

majormoses commented:
Thanks for your contribution to Sensu plugins! Without people like you submitting PRs we couldn't run the project. I will review it shortly.

majormoses commented on CHANGELOG.md:
```diff
@@ -4,6 +4,10 @@ This project adheres to [Semantic Versioning](http://semver.org/).
 
 ## [Unreleased]
+### Changed
+- check-mysql-replication-status: fix code flow if server is not a slave (@DrMurx)
```
this should technically go under `### Fixed` header but I can fix that up for you before I release it.

majormoses commented on CHANGELOG.md:
```diff
@@ -4,6 +4,10 @@ This project adheres to [Semantic Versioning](http://semver.org/).
+### Changed
+- check-mysql-replication-status: fix code flow if server is not a slave (@DrMurx)
+- check-mysql-replication-status: refactoring & spec tests (@DrMurx)
+- check-mysql-replication-status: added flapping protection (@DrMurx)
```
This should go under a `### Added` header but I can fix that up for you before I release it.

DrMurx commented:
> I will review the code in a minute but I did want to point out that sensu natively has its own flap detection algorithm borrowed from nagios. The Documentation is [here](https://docs.sensu.io/sensu-core/1.6/reference/checks/#check-attributes) look for `low_flap_threshold` and `high_flap_threshold`

Sensu's flapping protection is meant to prevent an unnecessary *storm* of alerting events in case a check is flapping. This PR does a completely different thing: it works around a condition of MySQL/MariaDB during which a replication slave incorrectly reports very high lag times (in my case it's >10 days!) for a short moment.

Of course I want any real lag to trigger a prompt alert, but I want the check to requery `SHOW SLAVE STATUS` under this specific condition.

DrMurx commented on bin/check-mysql-replication-status.rb:
```diff
@@ -122,64 +144,87 @@ def run
+    db = open_connection
+
+    retries = config[:flapping_retry]
+    # Note: Endless loop will exit via `ok`, `warning` or `critical`, or by Exception
```
This check is short-living. Any looping happens only in rare case we detect an unusually high replication lag and want to requery `SHOW SLAVE STATUS` to see whether it's a real lag or just a misreport by MySQL/MariaDB.

However, you're right, having an endless loop is some kind of anti-pattern, so I'll add a breaking condition, even if it's never met.

DrMurx commented on test/check-mysql-replication-status_spec.rb:
```diff
@@ -0,0 +1,113 @@
+#   check-mysql-replication-status_spec
+#
+# DESCRIPTION:
+#  rspec tests for check-mysql-replication-status
```
I see an issue with my choice of unit test - since the checkscript itself is `require`d by the test, the check will be executed once after the rspec tests have been finished and fail with `unknown` due to missing MySQL credentials.

I didn't come up with this pattern for a checkscript unit test myself, just borrowed it from another plugin's test. Any suggestions how to solve this?

majormoses replied:
I think this might be what you are looking for: https://github.com/sensu-plugins/sensu-plugins-rabbitmq/blob/7.0.0/test/spec_helper.rb#L7-L27

DrMurx commented:
@majormoses I'd like to go on with this one. With your experience in the plugin ecosystem, can you guide me to [a different way of testing without including the check script](https://github.com/sensu-plugins/sensu-plugins-mysql/pull/92#discussion_r243556203)? And maybe you have a [better proposal for naming](https://github.com/sensu-plugins/sensu-plugins-mysql/pull/92#issuecomment-449349129) the feature because "flapping" is already a standard term in monitoring solutions? As I wrote earlier, we're dealing with an bug/side effect of MySQL/MariaDB replication which rarely report outliers in the replication lag.

majormoses commented on bin/check-mysql-replication-status.rb:
```diff
@@ -93,6 +94,27 @@ class CheckMysqlReplicationStatus < Sensu::Plugin::Check::CLI
+         long: '--flapping-sleep=VALUE',
+         description: 'Sleep between flapping protection retries',
+         default: 1,
+         proc: lambda { |s| s.to_i } # rubocop:disable Lambda
```
Any particular reason to use a lambda over a proc that I am not seeing? This would remove the need to disable the cop.

```suggestion
         proc: proc { |s| s.to_i }
```

DrMurx replied:
Addressed in latest rework

majormoses commented:
I am having a hard time coming up with a good name as its pretty much trying to work around a reporting bug. Maybe something like `--discard-invalid-results-threshold`, `--discard-invalid-results-retries`, `--discard-invalid-results-sleep`? I can't say I like that either but nothing obvious is coming to mind.

DrMurx commented:
@majormoses I've picked up work on this PR again, and I renamed the feature to `lag outlier`; I think that wording expresses that it's an exceptional condition, yet unrelated to the flapping protection sensu already offers. 

Tests are now running properly. I removed the CodeClimate from `spec_helper.rb`, and added what you pointed out in comment https://github.com/sensu-plugins/sensu-plugins-mysql/pull/92#discussion_r258777728
